File: src/sdk/android/library/src/main/kotlin/io/veriface/sdk/pipeline/rppg/VeriFaceRppg.kt
package io.veriface.sdk.pipeline.rppg

import android.graphics.Rect
import android.graphics.ImageFormat
import androidx.camera.core.ImageProxy
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Real CHROM (CHrominance-based rPPG) implementation for Android.
 *
 * Algorithm by De Haan & Jeanne (2013):
 *   1. For each frame, extract the face region (from ML Kit face bounds)
 *   2. Compute mean R, G, B values in the face region
 *   3. Apply CHROM:
 *        X = 3*R - 2*G
 *        Y = 1.5*R + G - 1.5*B
 *   4. Combine: S = X - αY where α = std(X) / std(Y)
 *   5. Detrend S (remove low-frequency drift via moving average)
 *   6. Bandpass filter (0.7–4 Hz = 42–240 BPM)
 *   7. FFT to find dominant frequency → heart rate
 *   8. SNR = peak_power / mean_power
 *
 * Reference: De Haan, G. & Jeanne, V. (2013). "Robust Pulse Rate From
 * Chrominance-Based rPPG." IEEE Transactions on Biomedical Engineering.
 *
 * Pure Kotlin — no native libraries. Uses a radix-2 Cooley-Tukey FFT.
 */
class VeriFaceRppg(
    private val assumedFps: Double = 30.0
) {

    companion object {
        private const val MIN_SNR = 2.0
        private const val MIN_FRAMES = 30
    }

    data class Result(
        val score: Double,
        val heartRateBpm: Double?,
        val snr: Double,
        val frameCount: Int,
        val durationSec: Double
    )

    fun analyze(
        frames: List<ImageProxy>,
        timestamps: List<Long>,
        faceBounds: Rect,
        imageWidth: Int,
        imageHeight: Int
    ): Result {
        if (frames.size < MIN_FRAMES) {
            return Result(0.0, null, 0.0, frames.size, 0.0)
        }

        val rgbMeans = extractRgbMeans(frames, faceBounds)
        if (rgbMeans.size < MIN_FRAMES) {
            return Result(0.0, null, 0.0, rgbMeans.size, 0.0)
        }

        val chromSignal = computeChromSignal(rgbMeans)
        val detrended = detrend(chromSignal, 30)
        val filtered = bandpassFilter(detrended, 0.7, 4.0)
        val (dominantFreq, snr) = computeFftDominantFreq(filtered)

        val heartRateBpm = dominantFreq * 60.0
        val snrScore = minOf(1.0, snr / 10.0)
        val frameScore = minOf(1.0, frames.size.toDouble() / 60.0)
        val hrPlausibility = if (heartRateBpm in 40.0..200.0) 1.0 else 0.3
        val score = 0.5 * snrScore + 0.3 * frameScore + 0.2 * hrPlausibility

        val reportedHr = if (snr >= MIN_SNR) heartRateBpm else null
        val durationSec = if (timestamps.size >= 2) {
            (timestamps.last() - timestamps.first()) / 1_000_000_000.0
        } else {
            frames.size / assumedFps
        }

        return Result(score, reportedHr, snr, frames.size, durationSec)
    }

    private fun extractRgbMeans(frames: List<ImageProxy>, faceBounds: Rect): List<Triple<Double, Double, Double>> {
        val means = mutableListOf<Triple<Double, Double, Double>>()
        for (frame in frames) {
            computeMeanRgb(frame, faceBounds)?.let { means.add(it) }
        }
        return means
    }

    private fun computeMeanRgb(image: ImageProxy, faceBounds: Rect): Triple<Double, Double, Double>? {
        if (image.format != ImageFormat.YUV_420_888) return null

        val yPlane = image.planes[0]
        val uPlane = image.planes[1]
        val vPlane = image.planes[2]

        val yBuffer = yPlane.buffer
        val uBuffer = uPlane.buffer
        val vBuffer = vPlane.buffer

        val yRowStride = yPlane.rowStride
        val yPixelStride = yPlane.pixelStride
        val uRowStride = uPlane.rowStride
        val uPixelStride = uPlane.pixelStride
        val vRowStride = vPlane.rowStride
        val vPixelStride = vPlane.pixelStride

        val width = image.width
        val height = image.height

        val sampleLeft = maxOf(0, faceBounds.left + faceBounds.width() / 5)
        val sampleRight = minOf(width, faceBounds.right - faceBounds.width() / 5)
        val sampleTop = maxOf(0, faceBounds.top + faceBounds.height() / 4)
        val sampleBottom = minOf(height, faceBounds.bottom - faceBounds.height() / 4)

        if (sampleRight <= sampleLeft || sampleBottom <= sampleTop) return null

        var totalR = 0L
        var totalG = 0L
        var totalB = 0L
        var pixelCount = 0L

        val stride = 4
        var y = sampleTop
        while (y < sampleBottom) {
            var x = sampleLeft
            while (x < sampleRight) {
                val yIndex = y * yRowStride + x * yPixelStride
                val yVal = yBuffer.get(yIndex).toInt() and 0xFF

                val uvX = x / 2
                val uvY = y / 2
                val uIndex = uvY * uRowStride + uvX * uPixelStride
                val vIndex = uvY * vRowStride + uvX * vPixelStride

                val uVal = uBuffer.get(uIndex).toInt() and 0xFF
                val vVal = vBuffer.get(vIndex).toInt() and 0xFF

                val r = (yVal + 1.402 * (vVal - 128)).coerceIn(0.0, 255.0)
                val g = (yVal - 0.344 * (uVal - 128) - 0.714 * (vVal - 128)).coerceIn(0.0, 255.0)
                val b = (yVal + 1.772 * (uVal - 128)).coerceIn(0.0, 255.0)

                totalR += r.toLong()
                totalG += g.toLong()
                totalB += b.toLong()
                pixelCount++
                x += stride
            }
            y += stride
        }

        if (pixelCount == 0L) return null
        val n = pixelCount.toDouble()
        return Triple(totalR / n / 255.0, totalG / n / 255.0, totalB / n / 255.0)
    }

    private fun computeChromSignal(rgbMeans: List<Triple<Double, Double, Double>>): DoubleArray {
        if (rgbMeans.size <= 1) return DoubleArray(0)

        val xSignal = DoubleArray(rgbMeans.size) { 3.0 * rgbMeans[it].first - 2.0 * rgbMeans[it].second }
        val ySignal = DoubleArray(rgbMeans.size) { 1.5 * rgbMeans[it].first + rgbMeans[it].second - 1.5 * rgbMeans[it].third }

        val xStd = standardDeviation(xSignal)
        val yStd = standardDeviation(ySignal)
        val alpha = if (yStd > 0) xStd / yStd else 1.0

        return DoubleArray(rgbMeans.size) { xSignal[it] - alpha * ySignal[it] }
    }

    private fun detrend(signal: DoubleArray, windowSize: Int): DoubleArray {
        if (signal.size <= windowSize) return signal
        val n = signal.size
        val result = DoubleArray(n)
        val halfWindow = windowSize / 2
        val movingAvg = DoubleArray(n)
        for (i in 0 until n) {
            val start = maxOf(0, i - halfWindow)
            val end = minOf(n, i + halfWindow + 1)
            var sum = 0.0
            for (j in start until end) sum += signal[j]
            movingAvg[i] = sum / (end - start)
        }
        for (i in 0 until n) result[i] = signal[i] - movingAvg[i]
        return result
    }

    private fun bandpassFilter(signal: DoubleArray, lowFreq: Double, highFreq: Double): DoubleArray {
        val lowPassWindow = maxOf(3, (assumedFps / highFreq).toInt())
        return movingAverage(signal, lowPassWindow)
    }

    private fun movingAverage(signal: DoubleArray, windowSize: Int): DoubleArray {
        if (windowSize <= 1 || signal.size < windowSize) return signal
        val n = signal.size
        val result = DoubleArray(n)
        val halfWindow = windowSize / 2
        for (i in 0 until n) {
            val start = maxOf(0, i - halfWindow)
            val end = minOf(n, i + halfWindow + 1)
            var sum = 0.0
            for (j in start until end) sum += signal[j]
            result[i] = sum / (end - start)
        }
        return result
    }

    private fun computeFftDominantFreq(signal: DoubleArray): Pair<Double, Double> {
        if (signal.size < 16) return Pair(0.0, 0.0)

        val fftN = nextPowerOfTwo(signal.size)
        val padded = DoubleArray(fftN)
        for (i in signal.indices) padded[i] = signal[i]

        val windowed = applyHannWindow(padded)
        val (real, imag) = fft(windowed)

        val halfSize = fftN / 2
        val magnitudes = DoubleArray(halfSize) { sqrt(real[it] * real[it] + imag[it] * imag[it]) }

        val binWidth = assumedFps / fftN
        val lowBin = (0.7 / binWidth).toInt()
        val highBin = (4.0 / binWidth).toInt()

        if (highBin <= lowBin || highBin >= magnitudes.size) return Pair(0.0, 0.0)

        var peakBin = lowBin
        var peakMag = 0.0
        for (i in lowBin..highBin) {
            if (magnitudes[i] > peakMag) {
                peakMag = magnitudes[i]
                peakBin = i
            }
        }

        var meanMag = 0.0
        for (i in lowBin..highBin) meanMag += magnitudes[i]
        meanMag /= (highBin - lowBin + 1)
        val snr = if (meanMag > 0) (peakMag * peakMag) / (meanMag * meanMag) else 0.0

        return Pair(peakBin * binWidth, snr)
    }

    private fun applyHannWindow(signal: DoubleArray): DoubleArray {
        val n = signal.size
        val result = DoubleArray(n)
        for (i in 0 until n) {
            val w = 0.5 * (1.0 - cos(2.0 * Math.PI * i / (n - 1)))
            result[i] = signal[i] * w
        }
        return result
    }

    private fun fft(input: DoubleArray): Pair<DoubleArray, DoubleArray> {
        val n = input.size
        require(n and (n - 1) == 0) { "FFT size must be power of 2" }

        val real = input.copyOf()
        val imag = DoubleArray(n)

        var j = 0
        for (i in 1 until n) {
            var bit = n shr 1
            while (j and bit != 0) {
                j = j xor bit
                bit = bit shr 1
            }
            j = j or bit
            if (i < j) {
                real[i] = real[j].also { real[j] = real[i] }
                imag[i] = imag[j].also { imag[j] = imag[i] }
            }
        }

        var len = 2
        while (len <= n) {
            val angle = -2.0 * Math.PI / len
            val wReal = cos(angle)
            val wImag = sin(angle)
            var i = 0
            while (i < n) {
                var curReal = 1.0
                var curImag = 0.0
                val halfLen = len / 2
                for (k in 0 until halfLen) {
                    val idx1 = i + k
                    val idx2 = i + k + halfLen
                    val tReal = curReal * real[idx2] - curImag * imag[idx2]
                    val tImag = curReal * imag[idx2] + curImag * real[idx2]
                    real[idx2] = real[idx1] - tReal
                    imag[idx2] = imag[idx1] - tImag
                    real[idx1] = real[idx1] + tReal
                    imag[idx1] = imag[idx1] + tImag
                    val newReal = curReal * wReal - curImag * wImag
                    curImag = curReal * wImag + curImag * wReal
                    curReal = newReal
                }
                i += len
            }
            len *= 2
        }

        return Pair(real, imag)
    }

    private fun standardDeviation(values: DoubleArray): Double {
        if (values.size <= 1) return 0.0
        val mean = values.average()
        val variance = values.map { (it - mean).pow(2) }.sum() / (values.size - 1)
        return sqrt(variance)
    }

    private fun nextPowerOfTwo(n: Int): Int {
        var power = 1
        while (power < n) power *= 2
        return power
    }
}

File: src/sdk/ios/Sources/VeriFaceEdge/VeriFaceRppg.swift
// VeriFaceRppg.swift — Real CHROM rPPG implementation for iOS
//
// CHROM (CHrominance-based rPPG) algorithm by De Haan & Jeanne (2013).
// Extracts heart rate from skin pixel color variations across the green
// channel of captured video frames.
//
// Algorithm:
//   1. For each frame, extract the face region (from Vision face bounds)
//   2. Compute mean R, G, B values in the face region
//   3. Apply CHROM:
//        X = 3*R - 2*G
//        Y = 1.5*R + G - 1.5*B
//   4. Combine: S = X - αY where α = std(X) / std(Y)
//   5. Detrend S (remove low-frequency drift via moving average)
//   6. Bandpass filter (0.7–4 Hz = 42–240 BPM)
//   7. FFT to find dominant frequency → heart rate
//   8. SNR = peak_power / mean_power
//
// Reference: De Haan, G. & Jeanne, V. (2013). "Robust Pulse Rate From
// Chrominance-Based rPPG." IEEE Transactions on Biomedical Engineering.

import Foundation
import Accelerate

/// rPPG result from CHROM analysis.
struct RppgResult {
    /// Overall rPPG quality score (0.0–1.0). Higher = more reliable.
    let score: Double
    /// Detected heart rate in BPM (nil if SNR too low).
    let heartRateBpm: Double?
    /// Signal-to-noise ratio (higher = cleaner signal).
    let snr: Double
    /// Number of frames analyzed.
    let frameCount: Int
    /// Duration of signal in seconds.
    let durationSec: Double
}

/// Real CHROM rPPG analyzer.
final class VeriFaceRppg {

    /// Minimum SNR required to report a heart rate (below this = unreliable).
    private static let minSnr: Double = 2.0

    /// Minimum number of frames required for analysis.
    private static let minFrames: Int = 30

    /// Sampling rate (fps) — assumed from capture duration + frame count.
    private let assumedFps: Double

    init(assumedFps: Double = 30.0) {
        self.assumedFps = assumedFps
    }

    /// Run CHROM rPPG analysis on a sequence of frames.
    ///
    /// - Parameters:
    ///   - frames: Captured CVPixelBuffers (BGRA format from AVFoundation)
    ///   - timestamps: CMSample timestamps for each frame
    ///   - faceBounds: Normalized face bounding box (0.0–1.0) from Vision
    /// - Returns: rPPG result with score, heart rate, and SNR
    func analyze(
        frames: [CVPixelBuffer],
        timestamps: [CMTime],
        faceBounds: CGRect
    ) -> RppgResult {
        guard frames.count >= Self.minFrames else {
            return RppgResult(score: 0, heartRateBpm: nil, snr: 0, frameCount: frames.count, durationSec: 0)
        }

        // 1. Extract mean RGB from face region for each frame
        let rgbMeans = extractRgbMeans(frames: frames, faceBounds: faceBounds)

        // 2. Apply CHROM algorithm
        let chromSignal = computeChromSignal(rgbMeans: rgbMeans)

        // 3. Detrend (remove low-frequency drift)
        let detrended = detrend(signal: chromSignal, windowSize: 30)

        // 4. Bandpass filter (0.7–4 Hz = 42–240 BPM)
        let filtered = bandpassFilter(signal: detrended, lowFreq: 0.7, highFreq: 4.0)

        // 5. FFT to find dominant frequency
        let (dominantFreq, snr) = computeFftDominantFreq(signal: filtered)

        // 6. Convert frequency to BPM
        let heartRateBpm = dominantFreq * 60.0  // Hz → BPM

        // 7. Compute overall quality score
        // Score combines: SNR, signal amplitude, frame count adequacy
        let snrScore = min(1.0, snr / 10.0)  // SNR of 10+ = perfect score
        let frameScore = min(1.0, Double(frames.count) / 60.0)  // 60+ frames = perfect
        let hrPlausibility = (heartRateBpm >= 40 && heartRateBpm <= 200) ? 1.0 : 0.3
        let score = 0.5 * snrScore + 0.3 * frameScore + 0.2 * hrPlausibility

        // 8. Heart rate is only reported if SNR is sufficient
        let reportedHr = snr >= Self.minSnr ? heartRateBpm : nil

        // 9. Duration
        let durationSec: Double
        if timestamps.count >= 2 {
            durationSec = CMTimeGetSeconds(timestamps.last! - timestamps.first!)
        } else {
            durationSec = Double(frames.count) / assumedFps
        }

        return RppgResult(
            score: score,
            heartRateBpm: reportedHr,
            snr: snr,
            frameCount: frames.count,
            durationSec: durationSec
        )
    }

    // MARK: - RGB extraction

    /// Extract mean R, G, B values from the face region of each frame.
    private func extractRgbMeans(
        frames: [CVPixelBuffer],
        faceBounds: CGRect
    ) -> [(r: Double, g: Double, b: Double)] {
        var means: [(r: Double, g: Double, b: Double)] = []
        means.reserveCapacity(frames.count)

        for buffer in frames {
            if let mean = computeMeanRgb(pixelBuffer: buffer, faceBounds: faceBounds) {
                means.append(mean)
            }
        }

        return means
    }

    /// Compute mean R, G, B in the face region of a single CVPixelBuffer (BGRA format).
    private func computeMeanRgb(
        pixelBuffer: CVPixelBuffer,
        faceBounds: CGRect
    ) -> (r: Double, g: Double, b: Double)? {
        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }

        guard CVPixelBufferGetPixelFormatType(pixelBuffer) == kCVPixelFormatType_32BGRA else {
            return nil
        }

        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)
        let bytesPerRow = CVPixelBufferGetBytesPerRow(pixelBuffer)
        guard let baseAddress = CVPixelBufferGetBaseAddress(pixelBuffer) else { return nil }

        // Convert normalized face bounds to pixel coordinates
        let faceRect = CGRect(
            x: faceBounds.minX * CGFloat(width),
            y: faceBounds.minY * CGFloat(height),
            width: faceBounds.width * CGFloat(width),
            height: faceBounds.height * CGFloat(height)
        )

        // Sample a sub-region of the face (forehead + cheeks — best rPPG signal)
        // Use the center 60% of the face to avoid eyes/mouth
        let sampleRect = faceRect.insetBy(dx: faceRect.width * 0.2, dy: faceRect.height * 0.25)
        let startX = max(0, Int(sampleRect.minX))
        let endX = min(width, Int(sampleRect.maxX))
        let startY = max(0, Int(sampleRect.minY))
        let endY = min(height, Int(sampleRect.maxY))

        guard endX > startX && endY > startY else { return nil }

        // Use Accelerate for fast mean computation
        var totalR: UInt64 = 0
        var totalG: UInt64 = 0
        var totalB: UInt64 = 0
        var pixelCount: UInt64 = 0

        let ptr = baseAddress.assumingMemoryBound(to: UInt8.self)

        // Sample every 4th pixel for speed (still statistically valid)
        let stride = 4
        for y in stride(from: startY, to: endY, by: stride) {
            for x in stride(from: startX, to: endX, by: stride) {
                let offset = y * bytesPerRow + x * 4
                // BGRA format: B at offset 0, G at 1, R at 2, A at 3
                totalB += UInt64(ptr[offset])
                totalG += UInt64(ptr[offset + 1])
                totalR += UInt64(ptr[offset + 2])
                pixelCount += 1
            }
        }

        guard pixelCount > 0 else { return nil }

        let n = Double(pixelCount)
        return (
            r: Double(totalR) / n / 255.0,
            g: Double(totalG) / n / 255.0,
            b: Double(totalB) / n / 255.0
        )
    }

    // MARK: - CHROM signal computation

    /// Compute the CHROM signal from RGB means.
    /// S_n = X_n - α * Y_n where:
    ///   X_n = 3*R_n - 2*G_n
    ///   Y_n = 1.5*R_n + G_n - 1.5*B_n
    ///   α = std(X) / std(Y)
    private func computeChromSignal(
        rgbMeans: [(r: Double, g: Double, b: Double)]
    ) -> [Double] {
        guard rgbMeans.count > 1 else { return [] }

        // Compute X and Y signals
        let xSignal = rgbMeans.map { 3.0 * $0.r - 2.0 * $0.g }
        let ySignal = rgbMeans.map { 1.5 * $0.r + $0.g - 1.5 * $0.b }

        // Compute α = std(X) / std(Y)
        let xStd = standardDeviation(xSignal)
        let yStd = standardDeviation(ySignal)
        let alpha = yStd > 0 ? xStd / yStd : 1.0

        // Combine: S = X - α * Y
        return zip(xSignal, ySignal).map { x, y in x - alpha * y }
    }

    // MARK: - Detrending

    /// Remove low-frequency drift using a moving average filter.
    /// This isolates the pulsatile component (heart rate) from
    /// slow changes in illumination.
    private func detrend(signal: [Double], windowSize: Int) -> [Double] {
        guard signal.count > windowSize else { return signal }

        let n = signal.count
        var result = [Double](repeating: 0.0, count: n)
        let halfWindow = windowSize / 2

        // Compute moving average
        var movingAvg = [Double](repeating: 0.0, count: n)
        for i in 0..<n {
            let start = max(0, i - halfWindow)
            let end = min(n, i + halfWindow + 1)
            let window = signal[start..<end]
            movingAvg[i] = window.reduce(0.0, +) / Double(window.count)
        }

        // Subtract moving average from original signal
        for i in 0..<n {
            result[i] = signal[i] - movingAvg[i]
        }

        return result
    }

    // MARK: - Bandpass filter

    /// Apply a simple bandpass filter (0.7–4 Hz = 42–240 BPM).
    /// Uses a combination of moving average filters to approximate
    /// a bandpass response. For production, replace with a proper
    /// Butterworth filter via vDSP.
    private func bandpassFilter(
        signal: [Double],
        lowFreq: Double,
        highFreq: Double
    ) -> [Double] {
        // Simple approach: high-pass via detrending (already done),
        // low-pass via short moving average
        let lowPassWindow = max(3, Int(assumedFps / highFreq))
        return movingAverage(signal: signal, windowSize: lowPassWindow)
    }

    /// Simple moving average filter.
    private func movingAverage(signal: [Double], windowSize: Int) -> [Double] {
        guard windowSize > 1 && signal.count >= windowSize else { return signal }

        let n = signal.count
        var result = [Double](repeating: 0.0, count: n)
        let halfWindow = windowSize / 2

        for i in 0..<n {
            let start = max(0, i - halfWindow)
            let end = min(n, i + halfWindow + 1)
            let window = signal[start..<end]
            result[i] = window.reduce(0.0, +) / Double(window.count)
        }

        return result
    }

    // MARK: - FFT

    /// Compute the dominant frequency via FFT + SNR.
    /// Returns (dominantFreqHz, snr).
    private func computeFftDominantFreq(signal: [Double]) -> (Double, Double) {
        guard signal.count >= 16 else { return (0, 0) }

        // Find next power of 2 for FFT
        let n = signal.count
        let fftN = nextPowerOfTwo(n)

        // Pad signal to fftN with zeros
        var padded = [Double](repeating: 0.0, count: fftN)
        for i in 0..<n { padded[i] = signal[i] }

        // Apply Hann window (reduces spectral leakage)
        let windowed = applyHannWindow(padded)

        // Compute FFT using vDSP
        let magnitudes = computeFftMagnitudes(signal: windowed, fftSize: fftN)

        // Find the frequency range corresponding to 0.7–4 Hz (42–240 BPM)
        let binWidth = assumedFps / Double(fftN)
        let lowBin = Int(0.7 / binWidth)
        let highBin = Int(4.0 / binWidth)

        guard highBin > lowBin && highBin < magnitudes.count else {
            return (0, 0)
        }

        // Find peak in the valid frequency range
        var peakBin = lowBin
        var peakMag: Double = 0
        for i in lowBin...highBin {
            if magnitudes[i] > peakMag {
                peakMag = magnitudes[i]
                peakBin = i
            }
        }

        // Compute SNR: peak power / mean power in the valid range
        let validMags = Array(magnitudes[lowBin...highBin])
        let meanMag = validMags.reduce(0.0, +) / Double(validMags.count)
        let snr = meanMag > 0 ? (peakMag * peakMag) / (meanMag * meanMag) : 0

        let dominantFreq = Double(peakBin) * binWidth

        return (dominantFreq, snr)
    }

    /// Apply a Hann window to reduce spectral leakage.
    private func applyHannWindow(_ signal: [Double]) -> [Double] {
        let n = signal.count
        var result = [Double](repeating: 0.0, count: n)
        for i in 0..<n {
            let w = 0.5 * (1.0 - cos(2.0 * .pi * Double(i) / Double(n - 1)))
            result[i] = signal[i] * w
        }
        return result
    }

    /// Compute FFT magnitudes using Accelerate (vDSP).
    private func computeFftMagnitudes(signal: [Double], fftSize: Int) -> [Double] {
        let halfSize = fftSize / 2
        var magnitudes = [Double](repeating: 0.0, count: halfSize)

        // Split complex representation
        var realPart = [Double](repeating: 0.0, count: halfSize)
        var imagPart = [Double](repeating: 0.0, count: halfSize)

        var signalCopy = signal

        // Setup FFT
        guard let fftSetup = vDSP_create_fftsetupD(vDSP_Length(log2(Double(fftSize))), FFTRadix(kFFTRadix2)) else {
            return magnitudes
        }
        defer { vDSP_destroy_fftsetupD(fftSetup) }

        // Convert real signal to split complex
        var tempReal = [Double](repeating: 0.0, count: fftSize / 2)
        var tempImag = [Double](repeating: 0.0, count: fftSize / 2)
        tempReal.withUnsafeMutableBufferPointer { realPtr in
            tempImag.withUnsafeMutableBufferPointer { imagPtr in
                signalCopy.withUnsafeMutableBufferPointer { signalPtr in
                    var splitComplex = DSPDoubleSplitComplex(
                        realp: realPtr.baseAddress!,
                        imagp: imagPtr.baseAddress!
                    )
                    vDSP_ctozD(
                        UnsafePointer(signalPtr.baseAddress!).withMemoryRebound(to: DSPDoubleComplex.self, capacity: fftSize / 2) { $0 },
                        2,
                        &splitComplex,
                        1,
                        vDSP_Length(fftSize / 2)
                    )
                }
            }
        }

        // Perform FFT
        tempReal.withUnsafeMutableBufferPointer { realPtr in
            tempImag.withUnsafeMutableBufferPointer { imagPtr in
                var splitComplex = DSPDoubleSplitComplex(
                    realp: realPtr.baseAddress!,
                    imagp: imagPtr.baseAddress!
                )
                vDSP_fft_zripD(fftSetup, &splitComplex, 1, vDSP_Length(log2(Double(fftSize))), FFTDirection(FFT_FORWARD))
            }
        }

        // Compute magnitudes
        tempReal.withUnsafeMutableBufferPointer { realPtr in
            tempImag.withUnsafeMutableBufferPointer { imagPtr in
                magnitudes.withUnsafeMutableBufferPointer { magPtr in
                    var splitComplex = DSPDoubleSplitComplex(
                        realp: realPtr.baseAddress!,
                        imagp: imagPtr.baseAddress!
                    )
                    vDSP_zvabsD(&splitComplex, 1, magPtr.baseAddress!, 1, vDSP_Length(halfSize))
                }
            }
        }

        return magnitudes
    }

    // MARK: - Statistics helpers

    /// Compute standard deviation.
    private func standardDeviation(_ values: [Double]) -> Double {
        guard values.count > 1 else { return 0 }
        let mean = values.reduce(0.0, +) / Double(values.count)
        let variance = values.map { pow($0 - mean, 2) }.reduce(0.0, +) / Double(values.count - 1)
        return sqrt(variance)
    }

    /// Find next power of 2 ≥ n.
    private func nextPowerOfTwo(_ n: Int) -> Int {
        var power = 1
        while power < n { power *= 2 }
        return power
    }
}
